### ISSUE TITLE
Coinbase now returns Error reference number: 502 -- cloudflare" content="::CAPTCHA_BOX:: ::IM_UNDER_ATTACK_BOX::

### ISSUE THREAD

**Issue by ryny24** [ccxt/ccxt#19636]
### Operating System

Debian

### Programming Languages

PHP

### CCXT Version

4.1.21

### Description

I've been using ccxt for years, with a PHP script to monitor my orders every 10 minutes. It has worked great. Coinbase, which has been working for years, just started giving me this error about 1-2 days ago.  It appears that Cloudflare doesn't like my ccxt request every 10 minutes. If I stop my script for a while, it works again once and then starts the error again.

I was able to fix it, after finding a post about using the CURLOPT_COOKIEJAR option in the exchange setup, and copying my cookie from my desktop session. But then I realized, my VPS hosting provider has my Coinbase cookie!!!!! So that won't work.

Has anyone had this issue? I tried a cookie from an incognito session from Coinbase.com, not associated with my account but this gives me the same error. I also tried creating a dummy coinbase account, using that cookie. But no dice.

Thank you!

### Code

```
  fetch_balance from coinbase failed due to a network error: coinbase coinbase https://api.coinbase.com/api/v3/brokerage/products GET 502
<!DOCTYPE html>
<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
  <meta charset="utf-8">
  <title>Coinbase Error</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="cloudflare" content="::CAPTCHA_BOX:: ::IM_UNDER_ATTACK_BOX:: <div class="cf-error-details cf-error-502">
  <h1>Bad gateway</h1>
  <p>The web server reported a bad gateway error.</p>
  <ul>
    <li>Ray ID: xxxxxxxxxxxx</li>
    <li>Your IP address: x.x.x.x</li>
    <li>Error reference number: 502</li>
    <li>Cloudflare Location: Kansas City</li>
  </ul>
</div>


```

**Comment by carlosmiei** [comment]
Hello @ryny24,
Hum, actually I'm also getting a 502 error when accessing it 🤔 

![image](https://github.com/ccxt/ccxt/assets/43336371/8239b33a-dee6-4773-8786-1faab25728d9)

**Comment by carlosmiei** [comment]
@ryny24 yeah, I'm getting the same thing as you do. Maybe that's a temporary issue?
```
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="cloudflare" content="::CAPTCHA_BOX:: ::IM_UNDER_ATTACK_BOX:: <div class="cf-error-details cf-error-502">
  <h1>Bad gateway</h1>
  <p>The web server reported a bad gateway error.</p>
  <ul>
    <li>Ray ID: 81a0ffb638227744</li>
    <li>Your IP address: 185.44.76.187</li>
    <li>Error reference number: 502</li>
    <li>Cloudflare Location: London</li>
  </ul>
</div>
```

**Comment by ryny24** [comment]
So, doing a Google search for "cloudflare :IM_UNDER_ATTACK_BOX" brings up pages referring to customizing the Cloudflare error messages, all the way back to 2019. So the error is not new, but something must have changed recently.

Also, I believe my statement that CURLOPT_COOKIEJAR  can fix the issue was wrong. The issue is intermittent, all my attempts to fix it have failed.

**Comment by wylis-hodor** [comment]
FYI - I'm getting 502s right now using my own C program that uses libcurl.

**Comment by carlosmiei** [comment]
@ryny24 @wylis-hodor Can you test it again? It seems to be working now

**Comment by ryny24** [comment]
> @ryny24 @wylis-hodor Can you test it again? It seems to be working now

Still a problem here, I last got the error 4 minutes ago. (8:41 AM EST).  So far today, I got the error 30 times, but successfully returned my orders 24 times.

grep -c IM_UNDER_ATTACK_BOX 20231023.log
30

**Comment by madelmann** [comment]
I get the same error while being on CCXT 4.1.23 and on 4.1.18, so I guess it has nothing to do directly with the most recent changes in CCXT itself. I'm using Slang (a scripting language) and a somewhat recent libCurl extension.

**Comment by ryny24** [comment]
`Potential issues
Since the Under Attack mode requires your browser to support JavaScript to display and pass the interstitial page, it is expected to observe impact on third party analytics tools`

It seems to me, Coinbase enabled this setting (probably site wide) for some reason, and left it on. Not aware it can cause intermittent issues with the API. Hopefully they disable it soon.

**Comment by ryny24** [comment]
Seems like Coinbase has resolved the issue, I haven't seen the error in over 24 hours.

**Comment by madelmann** [comment]
Seems so.. I now get a different error:

{ "error": "INVALID_ARGUMENT", "error_details": "account is not available", "message": "account is not available" }

**Comment by GittyGitGitGitty** [comment]
I'm also getting something similar now: `{"errors":[{"id":"invalid_token","message":"The access token is invalid"}]}` but my token is valid. Generating a new token produces the same error.

**Comment by madelmann** [comment]
I've just checked out an older version of CCXT (4.1.43) and everything is working again for me, so I'
m quite sure that it's not an error on my side. Actually the issue is not only related to Coinbase but also the Bybit implementation is affected (all functions, not only fetching currencies). I will now gradually update to newer versions and update this entry when I've tracked down the version where the bug has been introduced.

**Comment by tyranosurasmax** [comment]
I'm getting a 502 error just by using Contracting.info.

**Comment by carlotestor** [comment]
Wrapping this one up since it's long resolved: the 502 / `IM_UNDER_ATTACK_BOX` page was Coinbase temporarily enabling Cloudflare "Under Attack" mode in front of `api.coinbase.com` — a server-side change, nothing in CCXT or your PHP could fix it, and it cleared on their end (as you confirmed).

For the record, cookies are never the answer here — API traffic isn't supposed to pass an interstitial, and copying a browser session cookie to a VPS is a bad idea (you were right to back out of it).

The durable handling for this class of outage is to catch it and retry, since CCXT maps Cloudflare pages to network errors:

```php
try {
    $orders = $exchange->fetch_open_orders ();
} catch (\ccxt\DDoSProtection $e) {
    // Cloudflare interstitial / 429 — back off and retry
    sleep (30);
} catch (\ccxt\ExchangeNotAvailable $e) {
    // 5xx from the exchange — retry later
    sleep (30);
}
```

The separate `invalid_token` / `account is not available` errors that appeared afterwards are unrelated — those come from Coinbase's migration to the Advanced Trade API; new-style CDP keys with `ccxt.coinbaseadvanced` (or `coinbase` on a current CCXT version) fix those.
